### PR TITLE
Contract tests fix (depends on #333)

### PR DESCRIPTION
**Pull request by dtebbs**
Fix and clean up some contract tests (#333)

- [x] Fix values and fully test residual bits
- [x] clean up variables
- [x] use a wrapping contract and rename to the actual contract being tested
- [x] remove dependency on prover server

(Does not move test logic into contract, but that may involve more complexity so maybe better addressed in the future)
Related to #335

### CODE REVIEW

File: zeth_contracts/contracts/MerkleTreeMiMC7_test.sol [hunk 1]
```diff
@@ -4,7 +4,8 @@
 
 pragma solidity ^0.5.0;
 
-import "./MerkleTreeMiMC7.sol";
+import "./BaseMerkleTree.sol";
```
Comment (AntoineRondelet) [inline zeth_contracts/contracts/MerkleTreeMiMC7_test.sol:7]: This is already fixed on #333

File: client/test_commands/test_altbn128_mixer_base.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+# SPDX-License-Identifier: LGPL-3.0+
+
+from zeth.core.constants import \
+    JS_INPUTS, ZETH_PUBLIC_UNIT_VALUE, ZETH_MERKLE_TREE_DEPTH
+import test_commands.mock as mock
+from typing import Any
+
+# pylint: disable=line-too-long
+
+# TODO: These test is specific to AltBN128MixerBase, however the mixer that is
```
Comment (AntoineRondelet) [inline client/test_commands/test_altbn128_mixer_base.py:14]: `These test is` -> `These tests are`

File: client/test_commands/test_altbn128_mixer_base.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    "0000000011101110111011101110111011101110111011101110111011101110",  # vout
+    2)
+
+PACKED_PRIMARY_INPUTS = \
+    [ROOT] + COMMITMENTS + NULLIFIERS + [HSIG] + HTAGS + [RESIDUAL_BITS]
+
+
+def test_assemble_nullifiers(mixer_instance: Any) -> None:
+    # Test retrieving nullifiers
+    print("--- testing ", "test_assemble_nullifiers")
```
Comment (AntoineRondelet) [inline client/test_commands/test_altbn128_mixer_base.py:97]: `test_assemble_nullifiers` -> `assemble_nullifier_test` (same for the functions below)

File: client/test_commands/test_altbn128_mixer_base.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    print("--- testing ", "test_assemble_hsig")
+    res = mixer_instance.functions.\
+        assemble_hsig_test(PACKED_PRIMARY_INPUTS).call()
+    hsig = int.from_bytes(res, byteorder="big")
+    assert hsig == HSIG, f"expected: {HSIG}, got {hsig}"
+
+
+def test_assemble_vpub(mixer_instance: Any) -> None:
+    # Test retrieving commitments
+    print("--- testing ", "test_assemble_vpub")
```
Comment (AntoineRondelet) [inline client/test_commands/test_altbn128_mixer_base.py:116]: `test_..._vpub` -> `assemble_public_values_test`

File: client/test_commands/test_altbn128_mixer_base.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    print("Deploying AltBN128MixerBase_test.sol")
+    _web3, eth = mock.open_test_web3()
+    deployer_eth_address = eth.accounts[0]
+    _mixer_interface, mixer_instance = mock.deploy_contract(
+        eth,
+        deployer_eth_address,
+        "AltBN128MixerBase_test",
+        {
+            'mk_depth': ZETH_MERKLE_TREE_DEPTH,
+        })
```
Comment (AntoineRondelet) [inline client/test_commands/test_altbn128_mixer_base.py:135]: What about having a linebreak before closing the function call here?

File: zeth_contracts/contracts/AltBN128MixerBase_test.sol [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+contract AltBN128MixerBase_test is AltBN128MixerBase
+{
+    constructor(uint256 mk_depth)
+        public
+        AltBN128MixerBase(mk_depth, address(0), new uint256[](0))
+    {
+    }
+
+    function assemble_public_values_test(uint256 residual_bits)
+        public pure
```
Comment (AntoineRondelet) [inline zeth_contracts/contracts/AltBN128MixerBase_test.sol:22]: I think `pure` should be on the next line

File: client/test_commands/test_altbn128_mixer_base.py (hunk 3 above)
Comment (dtebbs) [reply client/test_commands/test_altbn128_mixer_base.py:97]: Should this not be the name of the test? `assemble_nullifier_test` is the method name I happen to have used to wrap the internal methods.  Those solidity method names are more of an implementation detail (I was actually about to try reusing the same names - not sure if that works in solidity), whereas I guess the original `test_assemble_nullifiers` refers to the name of the test (the python function name).

Comment (dtebbs) [reply client/test_commands/test_altbn128_mixer_base.py:97]: I'll switch  it to: `print("--- test_assemble_nullifiers")`, which is hopefully a bit clearer.

File: client/test_commands/test_altbn128_mixer_base.py (hunk 5 above)
Comment (dtebbs) [reply client/test_commands/test_altbn128_mixer_base.py:135]: We could switch to that style, but I think most existing code keeps the closign parenthesis on the same line.

File: client/test_commands/test_altbn128_mixer_base.py (hunk 2 above)
Comment (dtebbs) [reply client/test_commands/test_altbn128_mixer_base.py:14]: fixed.  thanks.

File: client/test_commands/test_altbn128_mixer_base.py (hunk 3 above)
Comment (dtebbs) [reply client/test_commands/test_altbn128_mixer_base.py]: (changed to clearer print statements)

File: zeth_contracts/contracts/AltBN128MixerBase_test.sol (hunk 6 above)
Comment (dtebbs) [reply zeth_contracts/contracts/AltBN128MixerBase_test.sol]: moved to next line

File: client/test_commands/test_altbn128_mixer_base.py (hunk 4 above)
Comment (dtebbs) [reply client/test_commands/test_altbn128_mixer_base.py]: (as above.  `test_assemble_vpub` seems more suitable for printing.)

File: client/test_commands/test_altbn128_mixer_base.py (hunk 3 above)
Comment (AntoineRondelet) [reply client/test_commands/test_altbn128_mixer_base.py]: > I'll switch it to: print("--- test_assemble_nullifiers"), which is hopefully a bit clearer.

Thanks, that's a bit clearer (the `testing` prefix suggested that we ran a function named `test_assemble_nullifiers` which was not the case.).

> (changed to clearer print statements. Just FYI, I tried to keep the same names, but solidity does not seem to support overloading in this way, so I've kept the actual entry points as ...._test in the wrapper contract.)

Yes, let's keep the function names as they are If you changed the logs as per your comment above, that's fine. Thanks.

File: client/test_commands/test_altbn128_mixer_base.py (hunk 5 above)
Comment (AntoineRondelet) [reply client/test_commands/test_altbn128_mixer_base.py:135]: Ah right, let's stay consistent with existing code then.

File: zeth_contracts/contracts/MerkleTreeMiMC7_test.sol (hunk 1 above)
Comment (dtebbs) [reply zeth_contracts/contracts/MerkleTreeMiMC7_test.sol]: I've rebased onto the target branch.  thanks.

File: zeth_contracts/contracts/AltBN128MixerBase_test.sol [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        pure
+        returns(bytes32 nf)
+    {
+        return assemble_nullifier(index, primary_inputs);
+    }
+
+    // Dummy implementation of abstract function
+    function verify_zk_proof(
+        uint256[] memory proof,
+        uint256[NUM_INPUTS] memory inputs
```
Comment (AntoineRondelet) [inline zeth_contracts/contracts/AltBN128MixerBase_test.sol:53]: `proofs` and `inputs` are not used which makes the linter complain and causes the CI to fail. Let's either use them in a dummy condition, or switch the linter off for this block.

Comment (dtebbs) [reply zeth_contracts/contracts/AltBN128MixerBase_test.sol]: Using the `/* inputs */` approach, similar to C++ seems to satisfy the linter, if that works for you.

Comment (AntoineRondelet) [reply zeth_contracts/contracts/AltBN128MixerBase_test.sol]: Oh great, yes that's better.

**Review by AntoineRondelet — APPROVED** [review]
Thanks @dtebbs LGTM

**Pull request merged**